google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/GoogleCloudApiDetailsPanel.java

```diff
@@ -0,0 +1,259 @@
+import javax.swing.ImageIcon;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.JTextPane;
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import org.jetbrains.annotations.Nullable;
+
+/** The form-bound class for the Cloud API details panel. */
+public final class GoogleCloudApiDetailsPanel {
+
+  private static final String JAVA_CLIENT_LANGUAGE = "java";
+  private static final String LINKS_SEPARATOR = " | ";
+
+  private JLabel icon;
+  private JLabel nameLabel;
+  private JLabel versionLabel;
+  private JLabel statusLabel;
+  private JPanel panel;
+  private JTextPane descriptionTextPane;
+  private JTextPane linksTextPane;
+
+  private CloudLibrary currentCloudLibrary;
+
+  /** Returns the {@link JPanel} that holds the UI elements in this panel. */
+  JPanel getPanel() {
+    return panel;
+  }
+
+  /**
+   * Sets the currently displayed {@link CloudLibrary} to the given one.
+   *
+   * <p>If the given {@link CloudLibrary} is already displayed, nothing is done and no UI updates
+   * are made.
+   *
+   * @param library the {@link CloudLibrary} to display
+   */
+  void setCloudLibrary(CloudLibrary library) {
+    if (cloudLibrariesEqual(currentCloudLibrary, library)) {
+      return;
+    }
+    currentCloudLibrary = library;
+
+    if (library.getIcon() == null) {
+      icon.setIcon(null);
+    } else {
+      icon.setIcon(GoogleCloudToolsIcons.LOADING);
+      loadImageAsync(library.getIcon(), icon::setIcon);
+    }
+
+    nameLabel.setText(library.getName());
+    descriptionTextPane.setText(library.getDescription());
+
+    if (library.getClients() != null) {
+      library
+          .getClients()
+          .stream()
+          .filter(client -> JAVA_CLIENT_LANGUAGE.equals(client.getLanguage()))
+          .findFirst()
+          .ifPresent(client -> updateUI());
+    }
+  }
+
+  /**
+   * Initializes some UI components in this panel that require special set-up.
+   *
+   * <p>This is automatically called by the IDEA SDK and should not be directly invoked.
+   */
+  private void createUIComponents() {
+    descriptionTextPane = new JTextPane();
+    descriptionTextPane
+        .getDocument()
+        .addDocumentListener(
+            new DocumentListener() {
+              @Override
+              public void insertUpdate(DocumentEvent e) {
+                descriptionTextPane.setSize(
+                    descriptionTextPane.getWidth(), descriptionTextPane.getPreferredSize().height);
+              }
+
+              @Override
+              public void removeUpdate(DocumentEvent e) {}
+
+              @Override
+              public void changedUpdate(DocumentEvent e) {}
+            });
+
+    linksTextPane = new JTextPane();
+    linksTextPane.addHyperlinkListener(new BrowserOpeningHyperLinkListener());
+  }
+
+  /**
+   * Updates the UI elements in this panel to reflect the attributes in the {@link
+   * #currentCloudLibrary}.
+   */
+  private void updateUI() {
```

Comment:
this only updates the "client" portion of the UI right? the name, icon etc. is updated by the caller of this. Maybe the name / description can be updated to more clearly indicate this.

Reply:
Oh whoops, I forgot to remove the code above (in `.setCloudLibrary()`) when I added this method. The idea is that the setter should only set the `currentCloudLibrary` (if we need to) and call this method without doing any UI-related stuff, like it used to before this method.

PTAL, I've removed all the duplicate code in the calling method.